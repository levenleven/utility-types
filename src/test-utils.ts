/** @internal */
export function inferType<T, R extends T = T>(): T {
  return undefined as any;
}

// export interface IsType<T> {
//   (v: T): '1';
//   (v: any): '0';
// }
