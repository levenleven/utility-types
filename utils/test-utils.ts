/** @internal */
export function testType<T>(): T {
  return undefined as any;
}

// export interface IsType<T> {
//   (v: T): '1';
//   (v: any): '0';
// }
