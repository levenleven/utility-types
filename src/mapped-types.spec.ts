import { inferType } from './test-utils';
import {
  SetIntersection,
  SetDifference,
  SetComplement,
  SymmetricDifference,
  FunctionKeys,
  NonUndefined,
  NonFunctionKeys,
  Omit,
  Intersection,
  Diff,
  Subtract,
  Overwrite,
  Assign,
  Unionize,
  PromiseType,
  DeepReadonly,
  DeepRequired,
  DeepNonNullable,
} from '.';

/**
 * Fixtures
 */

type Props = { name: string; age: number; visible: boolean };
type DefaultProps = { age: number };
type NewProps = { age: string; other: string };
type MixedProps = { name: string; setName: (name: string) => void };

/**
 * Tests
 */
// TODO: make grouping using dts-jest
describe('mapped types', () => {
  it('SetIntersection', () => {
    // @dts-jest:pass:snap
    inferType<SetIntersection<'1' | '2' | '3', '2' | '3' | '4'>>();
    // @dts-jest:pass:snap
    inferType<SetIntersection<string | number | (() => void), () => void>>();
  });

  it('SetDifference', () => {
    // @dts-jest:pass:snap
    inferType<SetDifference<'1' | '2' | '3', '2' | '3' | '4'>>();
    // @dts-jest:pass:snap
    inferType<SetDifference<string | number | (() => void), () => void>>();
  });

  it('SetComplement', () => {
    // @dts-jest:pass:snap
    inferType<SetComplement<'1' | '2' | '3', '2' | '3'>>();
  });

  it('SymmetricDifference', () => {
    // @dts-jest:pass:snap
    inferType<SymmetricDifference<'1' | '2' | '3', '2' | '3' | '4'>>();
  });

  it('NonUndefined', () => {
    // @dts-jest:pass:snap
    inferType<NonUndefined<'1' | '2' | undefined>>();
    // @dts-jest:pass:snap
    inferType<NonUndefined<undefined>>();
  });

  it('FunctionKeys', () => {
    // @dts-jest:pass:snap
    inferType<FunctionKeys<MixedProps>>();
  });

  it('NonFunctionKeys', () => {
    // @dts-jest:pass:snap
    inferType<NonFunctionKeys<MixedProps>>();
  });

  it('Omit', () => {
    // @dts-jest:pass:snap
    inferType<Omit<Props, 'age'>>();
    // @dts-jest:pass:snap
    inferType<Omit<Props | NewProps, 'age'>>();
  });

  it('Intersection', () => {
    // @dts-jest:pass:snap
    inferType<Intersection<Props, DefaultProps>>();
    // @dts-jest:pass:snap
    inferType<Intersection<Props | NewProps, DefaultProps>>();
  });

  it('Diff', () => {
    // @dts-jest:pass:snap
    inferType<Diff<Props, NewProps>>();
  });

  it('Subtract', () => {
    // @dts-jest:pass:snap
    inferType<Subtract<Props, DefaultProps>>();
  });

  it('Overwrite', () => {
    // @dts-jest:pass:snap
    inferType<Overwrite<Props, NewProps>>();
  });

  it('Assign', () => {
    // @dts-jest:pass:snap
    inferType<Assign<Props, NewProps>>();
  });

  it('Unionize', () => {
    // @dts-jest:pass:snap
    inferType<Unionize<Props>>();
  });

  it('PromiseType', () => {
    // @dts-jest:pass:snap
    inferType<PromiseType<Promise<string>>>();
  });

  it('DeepReadonly', () => {
    type NestedProps = {
      first: {
        second: {
          name: string;
        };
      };
    };
    // @dts-jest:pass:snap
    inferType<DeepReadonly<NestedProps>['first']>();
    // @dts-jest:pass:snap
    inferType<DeepReadonly<NestedProps>['first']['second']>();
    // @dts-jest:pass:snap
    inferType<DeepReadonly<NestedProps>['first']['second']['name']>();

    type NestedArrayProps = {
      first: {
        second: Array<{ name: string }>;
      };
    };
    // @dts-jest:pass:snap
    inferType<DeepReadonly<NestedArrayProps>['first']>();
    // @dts-jest:pass:snap
    inferType<DeepReadonly<NestedArrayProps>['first']['second']>();
    // @dts-jest:pass:snap
    inferType<
      DeepReadonly<NestedArrayProps>['first']['second'][number]['name']
    >();
  });

  it('DeepRequired', () => {
    type NestedProps = {
      first?: {
        second?: {
          name?: string | null;
        };
      };
    };
    // @dts-jest:pass:snap
    inferType<DeepRequired<NestedProps>['first']>();
    // @dts-jest:pass:snap
    inferType<DeepRequired<NestedProps>['first']['second']>();
    // @dts-jest:pass:snap
    inferType<DeepRequired<NestedProps>['first']['second']['name']>();

    type NestedArrayProps = {
      first?: {
        second?: Array<{ name?: string | null } | undefined>;
      };
    };
    // @dts-jest:pass:snap
    inferType<DeepRequired<NestedArrayProps>['first']>();
    // @dts-jest:pass:snap
    inferType<DeepRequired<NestedArrayProps>['first']['second']>();
    // @dts-jest:pass:snap
    inferType<
      DeepRequired<NestedArrayProps>['first']['second'][number]['name']
    >();
  });

  it('DeepNonNullable', () => {
    type NestedProps = {
      first?: null | {
        second?: null | {
          name?: null | string;
        };
      };
    };
    // @dts-jest:pass:snap
    inferType<DeepNonNullable<NestedProps>['first']>();
    // @dts-jest:pass:snap
    inferType<DeepNonNullable<NestedProps>['first']['second']>();
    // @dts-jest:pass:snap
    inferType<DeepNonNullable<NestedProps>['first']['second']['name']>();

    type NestedArrayProps = {
      first?: null | {
        second?: Array<{ name?: string | null } | undefined | null>;
      };
    };
    // @dts-jest:pass:snap
    inferType<DeepNonNullable<NestedArrayProps>['first']>();
    // @dts-jest:pass:snap
    inferType<DeepNonNullable<NestedArrayProps>['first']['second']>();
    // @dts-jest:pass:snap
    inferType<
      DeepNonNullable<NestedArrayProps>['first']['second'][number]['name']
    >();
  });
});
