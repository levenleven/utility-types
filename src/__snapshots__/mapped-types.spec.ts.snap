// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`inferType<
    DeepNonNullable<NestedArrayProps>['first']['second'][number]['name']
  >() 1`] = `"string"`;

exports[`inferType<
    DeepReadonly<NestedArrayProps>['first']['second'][number]['name']
  >() 1`] = `"string"`;

exports[`inferType<
    DeepRequired<NestedArrayProps>['first']['second'][number]['name']
  >() 1`] = `"string | null"`;

exports[`inferType<Assign<Props, NewProps>>() 1`] = `"Pick<Pick<Props, \\"name\\" | \\"visible\\"> & Pick<NewProps, \\"age\\"> & Pick<NewProps, \\"other\\">, \\"name\\" | \\"age\\" | \\"visible\\" | \\"other\\">"`;

exports[`inferType<DeepNonNullable<NestedArrayProps>['first']>() 1`] = `"import(\\"/Users/piotrek/Dev/utility-types/src/mapped-types\\")._DeepNonNullableObject<{ second?: ({ name?: string | null | undefined; } | null | undefined)[] | undefined; }>"`;

exports[`inferType<DeepNonNullable<NestedArrayProps>['first']['second']>() 1`] = `"import(\\"/Users/piotrek/Dev/utility-types/src/mapped-types\\")._DeepNonNullableArray<{ name?: string | null | undefined; } | null | undefined>"`;

exports[`inferType<DeepNonNullable<NestedProps>['first']>() 1`] = `"import(\\"/Users/piotrek/Dev/utility-types/src/mapped-types\\")._DeepNonNullableObject<{ second?: { name?: string | null | undefined; } | null | undefined; }>"`;

exports[`inferType<DeepNonNullable<NestedProps>['first']['second']>() 1`] = `"import(\\"/Users/piotrek/Dev/utility-types/src/mapped-types\\")._DeepNonNullableObject<{ name?: string | null | undefined; }>"`;

exports[`inferType<DeepNonNullable<NestedProps>['first']['second']['name']>() 1`] = `"string"`;

exports[`inferType<DeepReadonly<NestedArrayProps>['first']>() 1`] = `"import(\\"/Users/piotrek/Dev/utility-types/src/mapped-types\\")._DeepReadonlyObject<{ second: { name: string; }[]; }>"`;

exports[`inferType<DeepReadonly<NestedArrayProps>['first']['second']>() 1`] = `"import(\\"/Users/piotrek/Dev/utility-types/src/mapped-types\\")._DeepReadonlyArray<{ name: string; }>"`;

exports[`inferType<DeepReadonly<NestedProps>['first']>() 1`] = `"import(\\"/Users/piotrek/Dev/utility-types/src/mapped-types\\")._DeepReadonlyObject<{ second: { name: string; }; }>"`;

exports[`inferType<DeepReadonly<NestedProps>['first']['second']>() 1`] = `"import(\\"/Users/piotrek/Dev/utility-types/src/mapped-types\\")._DeepReadonlyObject<{ name: string; }>"`;

exports[`inferType<DeepReadonly<NestedProps>['first']['second']['name']>() 1`] = `"string"`;

exports[`inferType<DeepRequired<NestedArrayProps>['first']>() 1`] = `"import(\\"/Users/piotrek/Dev/utility-types/src/mapped-types\\")._DeepRequiredObject<{ second?: ({ name?: string | null | undefined; } | undefined)[] | undefined; }>"`;

exports[`inferType<DeepRequired<NestedArrayProps>['first']['second']>() 1`] = `"import(\\"/Users/piotrek/Dev/utility-types/src/mapped-types\\")._DeepRequiredArray<{ name?: string | null | undefined; } | undefined>"`;

exports[`inferType<DeepRequired<NestedProps>['first']>() 1`] = `"import(\\"/Users/piotrek/Dev/utility-types/src/mapped-types\\")._DeepRequiredObject<{ second?: { name?: string | null | undefined; } | undefined; }>"`;

exports[`inferType<DeepRequired<NestedProps>['first']['second']>() 1`] = `"import(\\"/Users/piotrek/Dev/utility-types/src/mapped-types\\")._DeepRequiredObject<{ name?: string | null | undefined; }>"`;

exports[`inferType<DeepRequired<NestedProps>['first']['second']['name']>() 1`] = `"string | null"`;

exports[`inferType<Diff<Props, NewProps>>() 1`] = `"Pick<Props, \\"name\\" | \\"visible\\">"`;

exports[`inferType<FunctionKeys<MixedProps>>() 1`] = `"\\"setName\\""`;

exports[`inferType<Intersection<Props | NewProps, DefaultProps>>() 1`] = `"Pick<Props, \\"age\\"> | Pick<NewProps, \\"age\\">"`;

exports[`inferType<Intersection<Props, DefaultProps>>() 1`] = `"Pick<Props, \\"age\\">"`;

exports[`inferType<NonFunctionKeys<MixedProps>>() 1`] = `"\\"name\\""`;

exports[`inferType<NonUndefined<'1' | '2' | undefined>>() 1`] = `"\\"1\\" | \\"2\\""`;

exports[`inferType<NonUndefined<undefined>>() 1`] = `"never"`;

exports[`inferType<Omit<Props | NewProps, 'age'>>() 1`] = `"Pick<Props, \\"name\\" | \\"visible\\"> | Pick<NewProps, \\"other\\">"`;

exports[`inferType<Omit<Props, 'age'>>() 1`] = `"Pick<Props, \\"name\\" | \\"visible\\">"`;

exports[`inferType<Overwrite<Props, NewProps>>() 1`] = `"Pick<Pick<Props, \\"name\\" | \\"visible\\"> & Pick<NewProps, \\"age\\">, \\"name\\" | \\"age\\" | \\"visible\\">"`;

exports[`inferType<PromiseType<Promise<string>>>() 1`] = `"string"`;

exports[`inferType<SetComplement<'1' | '2' | '3', '2' | '3'>>() 1`] = `"\\"1\\""`;

exports[`inferType<SetDifference<'1' | '2' | '3', '2' | '3' | '4'>>() 1`] = `"\\"1\\""`;

exports[`inferType<SetDifference<string | number | (() => void), () => void>>() 1`] = `"string | number"`;

exports[`inferType<SetIntersection<'1' | '2' | '3', '2' | '3' | '4'>>() 1`] = `"\\"2\\" | \\"3\\""`;

exports[`inferType<SetIntersection<string | number | (() => void), () => void>>() 1`] = `"() => void"`;

exports[`inferType<Subtract<Props, DefaultProps>>() 1`] = `"Pick<Props, \\"name\\" | \\"visible\\">"`;

exports[`inferType<SymmetricDifference<'1' | '2' | '3', '2' | '3' | '4'>>() 1`] = `"\\"1\\" | \\"4\\""`;

exports[`inferType<Unionize<Props>>() 1`] = `"{ name: string; } | { age: number; } | { visible: boolean; }"`;
